File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        // approach: compare m to left sorted portion or right sorted portion
        
        int l = 0;
        int r = nums.length - 1;
        
        while (l <= r) {
            int m = (l + r) / 2;
            
            int mid = nums[m];
            int left = nums[l];
            int right = nums[r];
            
            if (mid == target) {
                return m;
            }
            
            if (left <= mid) { // left sorted portion
                if (target > mid || target < left) {
                    l = m + 1;
                } else {
                    r = m - 1;
                }
            } else { // right sorted portion
                if (target < mid || target > right) {
                    r = m - 1;
                } else {
                    l = m + 1;
                }
            }
        }
        
        return -1;
    }
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # approach: binary search while keeping track of left and right
        # sorted portions
        
        l = 0
        r = len(nums) - 1
        
        while l <= r:
            m = (l + r) // 2

            if nums[m] == target:
                return m
            
            if nums[m] >= nums[l]: # left sorted portion
                if target < nums[m] and target >= nums[l]:
                    # search left
                    r = m - 1
                else:
                    # search right
                    l = m + 1
            else: # right sorted portion
                if target > nums[m] and target <= nums[r]:
                    # search right
                    l = m + 1
                else:
                    # search left
                    r = m - 1
        return -1